### ISSUE TITLE
Annotate peaks with strandness 

### ISSUE THREAD

**Issue by crazyhottommy** [GuangchuangYu/ChIPseeker#17]
Hi GuangChuang,

Thanks for making such a nice tool. I am using ChIPseeker to annotate my genomic intervals (breakpoints from whole genome sequencing). Those are not ChIP-seq peaks and the strandness for each interval is critical to me.
6 column bed file format:

```
chr1  10000 100200  breakpoint_1  .   +
chr2  20000 200100 breakpoint_2  .    -
```

So, when annotating those intervals using

```R
annotatePeak(breakpointA, tssRegion=c(-3000, 3000), 
                         TxDb=txdb, annoDb="org.Hs.eg.db")
```

Is there a way to specify that strandness needs to be considered?
Note that, I am not saying the strandness of the genes (as you mentioned here http://guangchuangyu.github.io/2014/01/bug-of-r-package-chippeakanno/), but rather the strandness of the intervals.

Thanks,
Ming Tang

**Comment by GuangchuangYu** [comment]
Hi Ming,

Glad to know you are using `ChIPseeker`. Currently the strand info of intervals is not considered in `ChIPseeker`.  

Will let you know if this feature was supported.

Bests,
Guangchuang

**Comment by crazyhottommy** [comment]
Hi, 

If you are going to implement this. could you please make sure you read my questions here https://support.bioconductor.org/p/76512/ to get an idea of what I want and the suggesting solutions?

Thanks!
Ming

**Comment by GuangchuangYu** [comment]
```r
    require(GenomicRanges)
    require(ChIPseeker)
    f=getSampleFiles()[[4]]
    p=readPeakFile(f)
    p1 = p[1:6]

    ap1 = annotatePeak(p1)

    p2=p1
    strand(p2) = '+'

    ap2 = annotatePeak(p2)


    as.GRanges(ap1)

    ## GRanges object with 6 ranges and 11 metadata columns:
    ##       seqnames             ranges strand |          V4        V5
    ##          <Rle>          <IRanges>  <Rle> |    <factor> <numeric>
    ##   [1]     chr1 [ 815093,  817883]      * | MACS_peak_1    295.76
    ##   [2]     chr1 [1243288, 1244338]      * | MACS_peak_2     63.19
    ##   [3]     chr1 [2979977, 2981228]      * | MACS_peak_3    100.16
    ##   [4]     chr1 [3566182, 3567876]      * | MACS_peak_4    558.89
    ##   [5]     chr1 [3816546, 3818111]      * | MACS_peak_5     57.57
    ##   [6]     chr1 [6304865, 6305704]      * | MACS_peak_6     54.57
    ##             annotation  geneChr geneStart   geneEnd geneLength geneStrand
    ##            <character> <factor> <integer> <integer>  <integer>   <factor>
    ##   [1] Promoter (2-3kb)     chr1    803451    812182       8732          -
    ##   [2] Promoter (<=1kb)     chr1   1243994   1247057       3064          +
    ##   [3] Promoter (<=1kb)     chr1   2976181   2980350       4170          -
    ##   [4] Promoter (<=1kb)     chr1   3547331   3566671      19341          -
    ##   [5] Promoter (<=1kb)     chr1   3816968   3832011      15044          +
    ##   [6] Promoter (<=1kb)     chr1   6304252   6305638       1387          +
    ##            geneId transcriptId distanceToTSS
    ##       <character>  <character>     <numeric>
    ##   [1]      284593   uc001abt.4         -2911
    ##   [2]      126789   uc001aed.3             0
    ##   [3]      440556   uc001aka.3             0
    ##   [4]       49856   uc001ako.3             0
    ##   [5]   100133612   uc001alg.3             0
    ##   [6]      390992   uc009vly.2           613
    ##   -------
    ##   seqinfo: 24 sequences from hg19 genome

    as.GRanges(ap2)

    ## GRanges object with 6 ranges and 11 metadata columns:
    ##       seqnames             ranges strand |          V4        V5
    ##          <Rle>          <IRanges>  <Rle> |    <factor> <numeric>
    ##   [1]     chr1 [ 815093,  817883]      + | MACS_peak_1    295.76
    ##   [2]     chr1 [1243288, 1244338]      + | MACS_peak_2     63.19
    ##   [3]     chr1 [2979977, 2981228]      + | MACS_peak_3    100.16
    ##   [4]     chr1 [3566182, 3567876]      + | MACS_peak_4    558.89
    ##   [5]     chr1 [3816546, 3818111]      + | MACS_peak_5     57.57
    ##   [6]     chr1 [6304865, 6305704]      + | MACS_peak_6     54.57
    ##              annotation  geneChr geneStart   geneEnd geneLength geneStrand
    ##             <character> <factor> <integer> <integer>  <integer>   <factor>
    ##   [1]  Promoter (2-3kb)     chr1    803451    812182       8732          -
    ##   [2]  Promoter (<=1kb)     chr1   1243994   1247057       3064          +
    ##   [3] Distal Intergenic     chr1   2980636   2984289       3654          -
    ##   [4]  Promoter (1-2kb)     chr1   3569129   3652765      83637          +
    ##   [5]  Promoter (<=1kb)     chr1   3816968   3832011      15044          +
    ##   [6]  Promoter (<=1kb)     chr1   6304252   6305638       1387          +
    ##            geneId transcriptId distanceToTSS
    ##       <character>  <character>     <numeric>
    ##   [1]      284593   uc001abt.4         -2911
    ##   [2]      126789   uc001aed.3             0
    ##   [3]      440556   uc010nzg.1          3061
    ##   [4]        7161   uc001akp.3         -1253
    ##   [5]   100133612   uc001alg.3             0
    ##   [6]      390992   uc009vly.2           613
    ##   -------
    ##   seqinfo: 24 sequences from hg19 genome

    ap11 = annotatePeak(p1, ignoreDownstream=T)

    ap12 = annotatePeak(p1, ignoreUpstream=T)

    ap13 = annotatePeak(p1, ignoreOverlap=T)


    as.GRanges(ap11)

    ## GRanges object with 6 ranges and 11 metadata columns:
    ##       seqnames             ranges strand |          V4        V5
    ##          <Rle>          <IRanges>  <Rle> |    <factor> <numeric>
    ##   [1]     chr1 [ 815093,  817883]      * | MACS_peak_1    295.76
    ##   [2]     chr1 [1243288, 1244338]      * | MACS_peak_2     63.19
    ##   [3]     chr1 [2979977, 2981228]      * | MACS_peak_3    100.16
    ##   [4]     chr1 [3566182, 3567876]      * | MACS_peak_4    558.89
    ##   [5]     chr1 [3816546, 3818111]      * | MACS_peak_5     57.57
    ##   [6]     chr1 [6304865, 6305704]      * | MACS_peak_6     54.57
    ##             annotation  geneChr geneStart   geneEnd geneLength geneStrand
    ##            <character> <factor> <integer> <integer>  <integer>   <factor>
    ##   [1] Promoter (2-3kb)     chr1    803451    812182       8732          -
    ##   [2] Promoter (<=1kb)     chr1   1243994   1247057       3064          +
    ##   [3] Promoter (<=1kb)     chr1   2976181   2980350       4170          -
    ##   [4] Promoter (<=1kb)     chr1   3547331   3566671      19341          -
    ##   [5] Promoter (<=1kb)     chr1   3816968   3832011      15044          +
    ##   [6] Promoter (<=1kb)     chr1   6304252   6305638       1387          +
    ##            geneId transcriptId distanceToTSS
    ##       <character>  <character>     <numeric>
    ##   [1]      284593   uc001abt.4         -2911
    ##   [2]      126789   uc001aed.3             0
    ##   [3]      440556   uc001aka.3             0
    ##   [4]       49856   uc001ako.3             0
    ##   [5]   100133612   uc001alg.3             0
    ##   [6]      390992   uc009vly.2           613
    ##   -------
    ##   seqinfo: 24 sequences from hg19 genome

    as.GRanges(ap12)

    ## GRanges object with 6 ranges and 11 metadata columns:
    ##       seqnames             ranges strand |          V4        V5
    ##          <Rle>          <IRanges>  <Rle> |    <factor> <numeric>
    ##   [1]     chr1 [ 815093,  817883]      * | MACS_peak_1    295.76
    ##   [2]     chr1 [1243288, 1244338]      * | MACS_peak_2     63.19
    ##   [3]     chr1 [2979977, 2981228]      * | MACS_peak_3    100.16
    ##   [4]     chr1 [3566182, 3567876]      * | MACS_peak_4    558.89
    ##   [5]     chr1 [3816546, 3818111]      * | MACS_peak_5     57.57
    ##   [6]     chr1 [6304865, 6305704]      * | MACS_peak_6     54.57
    ##              annotation  geneChr geneStart   geneEnd geneLength geneStrand
    ##             <character> <factor> <integer> <integer>  <integer>   <factor>
    ##   [1] Distal Intergenic     chr1    852953    854817       1865          -
    ##   [2]  Promoter (<=1kb)     chr1   1243994   1247057       3064          +
    ##   [3]  Promoter (<=1kb)     chr1   2976181   2980350       4170          -
    ##   [4]  Promoter (<=1kb)     chr1   3547331   3566671      19341          -
    ##   [5]  Promoter (<=1kb)     chr1   3816968   3832011      15044          +
    ##   [6]            3' UTR     chr1   6307406   6321035      13630          -
    ##            geneId transcriptId distanceToTSS
    ##       <character>  <character>     <numeric>
    ##   [1]   100130417   uc010nxy.1         36934
    ##   [2]      126789   uc001aed.3             0
    ##   [3]      440556   uc001aka.3             0
    ##   [4]       49856   uc001ako.3             0
    ##   [5]   100133612   uc001alg.3             0
    ##   [6]      387509   uc001amp.2         15331
    ##   -------
    ##   seqinfo: 24 sequences from hg19 genome

    as.GRanges(ap13)

    ## GRanges object with 6 ranges and 11 metadata columns:
    ##       seqnames             ranges strand |          V4        V5
    ##          <Rle>          <IRanges>  <Rle> |    <factor> <numeric>
    ##   [1]     chr1 [ 815093,  817883]      * | MACS_peak_1    295.76
    ##   [2]     chr1 [1243288, 1244338]      * | MACS_peak_2     63.19
    ##   [3]     chr1 [2979977, 2981228]      * | MACS_peak_3    100.16
    ##   [4]     chr1 [3566182, 3567876]      * | MACS_peak_4    558.89
    ##   [5]     chr1 [3816546, 3818111]      * | MACS_peak_5     57.57
    ##   [6]     chr1 [6304865, 6305704]      * | MACS_peak_6     54.57
    ##             annotation  geneChr geneStart   geneEnd geneLength geneStrand
    ##            <character> <factor> <integer> <integer>  <integer>   <factor>
    ##   [1] Promoter (2-3kb)     chr1    803451    812182       8732          -
    ##   [2] Promoter (<=1kb)     chr1   1227764   1243269      15506          -
    ##   [3]           3' UTR     chr1   2980636   2984289       3654          -
    ##   [4] Promoter (1-2kb)     chr1   3569129   3652765      83637          +
    ##   [5]           5' UTR     chr1   3773845   3782564       8720          +
    ##   [6] Promoter (<=1kb)     chr1   6304252   6305638       1387          +
    ##            geneId transcriptId distanceToTSS
    ##       <character>  <character>     <numeric>
    ##   [1]      284593   uc001abt.4         -2911
    ##   [2]      116983   uc001aeb.2           -19
    ##   [3]      440556   uc010nzg.1          3061
    ##   [4]        7161   uc001akp.3         -1253
    ##   [5]        1677   uc009vln.1         42701
    ##   [6]      390992   uc009vly.2           613
    ##   -------
    ##   seqinfo: 24 sequences from hg19 genome
```

I update ChIPseeker to support strandness of genomic interval, and add additional parameters for only considering upstream/downstream.

Please help test it.

**Comment by crazyhottommy** [comment]
was going to test the latest version. bioc version is 1.6.4. tried:

```
devtools::install_github("GuangchuangYu/ChIPseeker")
Downloading GitHub repo GuangchuangYu/ChIPseeker@master
Installing ChIPseeker
Skipping 1 packages not available: TxDb.Hsapiens.UCSC.hg19.knownGene
Installing 3 packages: GenomicAlignments, GenomicRanges, S4Vectors
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.
Error: file ‘/var/folders/79/x06wz9v560q10gw881n9c_z0x7m0vl/T//RtmpQEOkoX/downloaded_packages/S4Vectors_0.8.6.tgz’ is not an OS X binary package
In addition: Warning message:
'tar' returned non-zero exit code 1 
```

**Comment by GuangchuangYu** [comment]
```r
pkgs = c("GenomicAlignments", "GenomicRanges", "S4Vectors")
install.packages(pkgs, repos = "http://www.bioconductor.org/packages/devel/bioc",
                          type = "source")
library(BiocInstaller)
biocLite("guangchuangyu/ChIPseeker")
```

**Comment by crazyhottommy** [comment]
Got errors when loading the library. 
```
 biocLite("guangchuangyu/ChIPseeker")
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.2 (BiocInstaller 1.20.1), R 3.2.2 (2015-08-14).
Installing github package(s) ‘guangchuangyu/ChIPseeker’
Downloading GitHub repo guangchuangyu/ChIPseeker@master
Installing ChIPseeker
Skipping 1 packages not available: TxDb.Hsapiens.UCSC.hg19.knownGene
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save --no-restore CMD  \
  INSTALL  \
  '/private/var/folders/79/x06wz9v560q10gw881n9c_z0x7m0vl/T/RtmpXQk4Yr/devtoolsc1e1c34a417/GuangchuangYu-ChIPseeker-d741c95'  \
  --library='/Users/mtang1/Library/R/3.2/library' --install-tests 

* installing *source* package ‘ChIPseeker’ ...
** R
** data
** inst
** preparing package for lazy loading
Warning: replacing previous import by ‘BiocGenerics::unstrand’ when loading ‘ChIPseeker’
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Warning: replacing previous import by ‘BiocGenerics::unstrand’ when loading ‘ChIPseeker’
* DONE (ChIPseeker)


> library(ChIPseeker)
Error in get(method, envir = home) : 
  lazy-load database '/Users/mtang1/Library/R/3.2/library/ChIPseeker/R/ChIPseeker.rdb' is corrupt
In addition: Warning messages:
1: replacing previous import by ‘BiocGenerics::unstrand’ when loading ‘ChIPseeker’ 
2: In .registerS3method(fin[i, 1], fin[i, 2], fin[i, 3], fin[i, 4],  :
  restarting interrupted promise evaluation
3: In get(method, envir = home) :
  restarting interrupted promise evaluation
4: In get(method, envir = home) : internal error -3 in R_decompress1
Error: package or namespace load failed for ‘ChIPseeker’

```

**Comment by GuangchuangYu** [comment]
try
```r
pkgs <- c("SummarizedExperiment", "BiocGenerics", "GenomicAlignments", "GenomicRanges", "S4Vectors", "GenomeInfoDb", "IRanges")
install.packages(pkgs, repos = "http://www.bioconductor.org/packages/devel/bioc",
                           type = "source")
biocLite("guangchuangyu/ChIPseeker")
```

I can guarantee it will work, since there are a lot of namespace change in devel branch of Bioc base packages (see the [commit](https://github.com/GuangchuangYu/ChIPseeker/commit/91f7af45020439d0fdf40d6781de17c051c8105b)).

**Comment by GuangchuangYu** [comment]
maybe you just need to re-run R after the installation, see <https://groups.google.com/forum/#!topic/h2ostream/IEQDf1dVw44>

**Comment by crazyhottommy** [comment]
thx Guangchuang! It is working now, and I manually checked several breakpoints and  `ignoreDownstream = TRUE` seems to be working as I expected! 

I will bug you when I have questions.

Best,
Ming

**Comment by crazyhottommy** [comment]
Hi Guangchuan,

I updated ChIPseeker to the latest version, and now it is giving me error.

```r
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

## ignoreDownstream will only find genes upstream of the breakpoint.
## overlap will find the genes with which the breakpoint overlaps in the gene body, not necessary TSS region.

breakA_anno<- annotatePeak(breakpointA, tssRegion=c(-3000, 3000), 
                         TxDb=txdb, level = "transcript", annoDb="org.Hs.eg.db",
                         sameStrand = FALSE, ignoreOverlap = FALSE,
                         ignoreDownstream = TRUE, overlap = "all")
>> preparing features information...		 2016-06-07 15:38:01 
>> identifying nearest features...		 2016-06-07 15:38:02 
Error: subscript contains NAs or out-of-bounds indices


> sessionInfo()
R version 3.3.0 (2016-05-03)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.11.5 (El Capitan)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] BSgenome.Hsapiens.UCSC.hg19_1.4.0       BSgenome_1.40.0                        
 [3] rtracklayer_1.32.0                      Biostrings_2.40.1                      
 [5] XVector_0.12.0                          ggplot2_2.1.0                          
 [7] tidyr_0.4.1                             dplyr_0.4.3                            
 [9] GenomicInteractions_1.6.2               InteractionSet_1.0.0                   
[11] SummarizedExperiment_1.2.2              TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
[13] GenomicFeatures_1.24.2                  AnnotationDbi_1.34.3                   
[15] Biobase_2.32.0                          GenomicRanges_1.24.1                   
[17] GenomeInfoDb_1.8.1                      IRanges_2.6.0                          
[19] S4Vectors_0.10.1                        BiocGenerics_0.18.0                    
[21] ChIPseeker_1.9.3                       

loaded via a namespace (and not attached):
 [1] httr_1.1.0                    AnnotationHub_2.4.2           splines_3.3.0                
 [4] gtools_3.5.0                  Formula_1.2-1                 shiny_0.13.2                 
 [7] assertthat_0.1                interactiveDisplayBase_1.10.3 latticeExtra_0.6-28          
[10] Rsamtools_1.24.0              RSQLite_1.0.0                 lattice_0.20-33              
[13] biovizBase_1.20.0             chron_2.3-47                  digest_0.6.9                 
[16] RColorBrewer_1.1-2            colorspace_1.2-6              htmltools_0.3.5              
[19] httpuv_1.3.3                  Matrix_1.2-6                  plyr_1.8.3                   
[22] XML_3.98-1.4                  biomaRt_2.28.0                zlibbioc_1.18.0              
[25] xtable_1.8-2                  scales_0.4.0                  gdata_2.17.0                 
[28] BiocParallel_1.6.2            UpSetR_1.2.1                  nnet_7.3-12                  
[31] Gviz_1.16.1                   survival_2.39-3               magrittr_1.5                 
[34] mime_0.4                      gplots_3.0.1                  foreign_0.8-66               
[37] BiocInstaller_1.22.2          tools_3.3.0                   data.table_1.9.6             
[40] matrixStats_0.50.2            gridBase_0.4-7                stringr_1.0.0                
[43] munsell_0.4.3                 cluster_2.0.4                 plotrix_3.6-2                
[46] ensembldb_1.4.5               caTools_1.17.1                grid_3.3.0                   
[49] RCurl_1.95-4.8                dichromat_2.0-0               VariantAnnotation_1.18.0     
[52] igraph_1.0.1                  bitops_1.0-6                  boot_1.3-18                  
[55] gtable_0.2.0                  DBI_0.4-1                     R6_2.1.2                     
[58] GenomicAlignments_1.8.1       gridExtra_2.2.1               Hmisc_3.17-4                 
[61] KernSmooth_2.23-15            stringi_1.1.1                 Rcpp_0.12.5                  
[64] rpart_4.1-10                  acepack_1.3-3.3 

```

I can send you breakpointA for debugging.

Thanks,
Ming

**Comment by GuangchuangYu** [comment]
This was introduced by <https://github.com/GuangchuangYu/ChIPseeker/pull/24/files>, that the `dummyID` is actually `length(features) + 1` which caused this out-of-bound issue.

This bug was fixed in >=1.8.4 (release) and >= 1.9.4 (devel).
